File: util-rs/src/z_algorithm.rs
// https://www.geeksforgeeks.org/z-algorithm-linear-time-pattern-searching-algorithm/
// Z-Arrayの構築
// O(N)
// for i < n, z[i] is max z[i] < n such that s[0:z[i]] = s[i:i+z[i]]
// パターンの検索にも使える
#[allow(dead_code)]
fn z_algorithm<T: std::cmp::Eq>(s: &Vec<T>) -> Vec<usize> {
    let n = s.len();

    // z[i] = max_{j<n} s[0:j] = s[i:i+j]
    let mut z = vec![0; n];
    z[0] = n;

    let mut l = 0;
    let mut r = 0;
    for i in 1..n {
        // assert!(s[l..r] == s[0..r - l]);
        if i < r && z[i - l] < r - i {
            z[i] = z[i - l];
        } else {
            // i < rなら、 z[i - l] >= r - i なので、
            // s[i..r] (=s[i-l..r-l]) = s[0..r-i] が保証されている
            // i >= r なら再計算
            l = i;
            r = std::cmp::max(i, r);
            while r < n && s[r] == s[r - l] {
                r += 1;
            }
            z[i] = r - l;
        }
    }

    z
}

// textからpatternの出現箇所を全部検索
#[allow(dead_code)]
fn find_all<T: Eq>(text: &Vec<T>, pattern: &Vec<T>) -> Vec<usize> {
    // pattern + text
    let s = pattern.iter().chain(text.iter()).collect::<Vec<_>>();

    let z = z_algorithm(&s);

    (0..text.len())
        .filter(|&i| z[pattern.len() + i] >= pattern.len())
        .collect()
}

#[test]
fn test_find_all() {
    let s = "ABC ABCDAB ABCDABCDABDEABCDABDFF"
        .chars()
        .collect::<Vec<_>>();
    let w = "ABCDABD".chars().collect::<Vec<_>>();

    assert_eq!(find_all(&s, &w), vec![15, 23]);
}

File: cpp/util-cpp/z_algorithm.cpp
#include <bits/stdc++.h>

#include <boost/range/irange.hpp>

using namespace std;
using namespace boost;

// https://www.geeksforgeeks.org/z-algorithm-linear-time-pattern-searching-algorithm/
// Z-Arrayの構築
// O(N)
// for i < n, z[i] is max z[i] < n such that s[0:z[i]] = s[i:i+z[i]]
// パターンの検索にも使える

template <typename S>
vector<int64_t> z_algorithm(const S& s) {
    int64_t n = s.size();

    // z[i] = max_{j<n} s[0:j] = s[i:i+j]
    vector<int64_t> z(n, 0L);
    z[0] = n;

    int64_t l = 0, r = 0;
    for (auto i : irange(1L, n)) {
        if (i < r && z[i - l] < r - i) {
            z[i] = z[i - l];
        } else {
            // i < rなら、 z[i - l] >= r - i なので、
            // s[i:r] (=s[i-l:r-l]) = s[0:r-i] が保証されている
            // i >= r なら再計算
            l = i;
            r = max(i, r);
            while (r < n && s[r] == s[r - l]) {
                ++r;
            }
            z[i] = r - l;
        }
    }

    return z;
}

template <typename S>
vector<int64_t> find_all(const S& text, const S& pattern) {
    auto s = pattern;
    s.push_back('\0');
    s += text;

    auto z = z_algorithm(s);

    vector<int64_t> result;
    for (auto i : irange(0uL, text.size())) {
        if (z[pattern.size() + 1 + i] == pattern.size()) {
            result.push_back(i);
        }
    }

    return result;
}

int main() {
    string s("ABC ABCDAB ABCDABCDABDEABCDABDFF");
    string w("ABCDABD");

    for (auto idx : find_all(s, w)) {
        cout << idx << endl;
    }
}